Working impl of HK2 dependency injection

collado-mike opened the pull request:
# Description

As discussed in the mailing list, this change proposes using the HK2 framework to support CDI-compliant dependency injection using the existing Dropwizard application structure and yaml configuration. This change does not make a logic changes or signature changes except to remove some of the injection-oriented interfaces (e.g., `ConfigurationStoreAware`) and updates constructors to accept dependencies via `@Inject` annotations.

The `PolarisApplicationConfig` is configured to be managed by HK2, but populated by the Jackson during the typical Dropwizard configuration parsing. Because the `PolarisApplicationConfig` is managed by HK2, it can invoke a `@PostConstruct` method that registers configured beans with the `serviceLocator` using the config class's own getter methods as `Factory` implementations. Because the other beans are _not_ managed by HK2 (they can't be, since they declare dependencies on other beans that may not yet be instantiated by Jackson), the factory manually calls `inject` on each bean before returning.

Jersey has its own `serviceLocator` instance under the hood, so all of the beans bound to the `PolarisApplicationConfig` instance are registered again with the Jersey `serviceLocator` so that they can be found by the jersey components at runtime (I tried using the HK2 bridge that supposedly allows connecting two `serviceLocator`s, but it didn't seem to work and also the licensing on that package suggests possible GPL violation).

This also introduces a `@RealmScope` that declares certain dependencies to be created and managed per-realm, rather than per-application or per-request. Some of the instances, such as the `PolarisMetaStoreManager` and the `EntityCache` are intended to be reused across requests, but of course, cannot be reused across realms, as some entities (such as `root` principal and `service_admin` principal role) have the same name across realms. 

One additional test class is added to ensure that the `@RealmScope`d `EntityCache` is reused across requests, but is not reused across different realms.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] PolarisRealmEntityCacheTest

**Test Configuration**:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

Please delete options that are not relevant.

- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] If adding new functionality, I have discussed my implementation with the community using the linked GitHub issue

dimas-b commented on polaris-service/src/main/java/org/apache/polaris/service/persistence/cache/EntityCacheFactory.java:
```diff
@@ -0,0 +1,47 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EntityCacheFactory implements Factory<EntityCache> {
```
Suggestion: `jakarta.enterprise.inject.Produces` from jakarta.enterprise.cdi-api 4.1.0

collado-mike replied:
HK2 doesn't implement support for `@Produces`. It does ship with its own `@Provides` annotation (see https://github.com/eclipse-ee4j/glassfish-hk2/pull/488 ), but that annotation is HK2-specific and doesn't really solve the intended goal, I think. It's possible to add something, but I think we should avoid cluttering this PR and leave that as an improvement to make afterward.

dimas-b replied:
Well, in this case it looks like factories still depend on HK2 classes heavily, which will complicate reuse (HK2 will trickle down all downstream runtime environments). I'm not sure what the benefit of using HK2 will be if we cannot make it a runtime-only dependency, but it effectively becomes an API dependency (meaning any attempt to customize DI will have to bind to HK2 APIs).

collado-mike replied:
The HK2 factory classes can be restricted to the dropwizard runtime. There are no references to these factory classes outside of the `PolarisApplication`, which is Dropwizard-specific

dimas-b commented on extension/persistence/eclipselink/build.gradle.kts:
```diff
@@ -32,6 +32,7 @@ dependencies {
   implementation(libs.eclipselink)
   implementation(platform(libs.dropwizard.bom))
   implementation("io.dropwizard:dropwizard-jackson")
+  implementation("org.glassfish.hk2:hk2-api:3.0.6")
```
Why does EclipseLink need an explicit dependency on HK2? Shouldn't is be sufficient to put HK2 in the DW application?

collado-mike replied:
This was from when I had the MetaStoreManagerFactory extend the Factory interface - it's unnecessary now. I'll remove it.

dimas-b replied:
it's still here? ;)

dimas-b reviewed (commented):
Just a couple of quick comments... I'll do a full review a bit later :)

snazy commented on polaris-core/build.gradle.kts:
```diff
@@ -72,6 +68,7 @@ dependencies {
   implementation(libs.swagger.annotations)
   implementation(libs.swagger.jaxrs)
   implementation(libs.jakarta.validation.api)
+  implementation("jakarta.inject:jakarta.inject-api:2.0.1")
```
Should be managed via the version catalog.

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/persistence/MetaStoreManagerFactory.java:
```diff
@@ -19,33 +19,26 @@
  * Configuration interface for configuring the {@link PolarisMetaStoreManager} via Dropwizard
  * configuration
  */
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
```
While this and #469 aim to migrate to "proper CDI", the use of DI via Jackson should be removed as well.

collado-mike replied:
Yeah, I think I missed a handful of these. I'll go back and revisit existing types.

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegrationProvider.java:
```diff
@@ -18,12 +18,17 @@
  * Factory interface that knows how to construct a {@link PolarisStorageIntegration} given a {@link
  * PolarisStorageConfigurationInfo}.
  */
+@JsonTypeInfo(
```
Same comment as above, let's not rely on Jackson for DI.

snazy commented on polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:
```diff
@@ -151,87 +173,195 @@ public void initialize(Bootstrap<PolarisApplicationConfig> bootstrap) {
+            new AbstractBinder() {
+              @Override
+              protected void configure() {
+                bind(RealmScopeContext.class)
```
Looks like there's a lot of manual coding needed for hk2, which feels like it makes things complicated, when we eventually have a plugin/extension mechanism.

collado-mike replied:
I think this should be considered the smallest change necessary to get CDI working. There are improvements that can be made, e.g., classpath scanning for detecting Resource and Service impls, but the primary goal right now is to have runtime framework dependent on `@Inject` for wiring. Once that's done, lots of other changes can happen in parallel - e.g., improvements to the authorizer, authentication changes, as well as improvements to the CDI framework. I think this accomplishes the smallest change necessary to allow those other changes to progress.

snazy commented on polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:
```diff
@@ -412,4 +533,31 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
     }
   }
+
+  private static class PolarisMetaStoreManagerFactory implements Factory<PolarisMetaStoreManager> {
```
Should be replaced with `@Produces` and `@Disposes`

collado-mike replied:
Mentioned above - HK2 doesn't support `@Produces` out of the box. HK2 implements "CDI-lite" whereas the enterprise.* annotations are defined in the "CDI-full" spec

snazy commented on polaris-service/src/main/java/org/apache/polaris/service/auth/DefaultOAuth2ApiService.java:
```diff
@@ -39,8 +40,8 @@
  */
-@JsonTypeName("default")
-public class DefaultOAuth2ApiService implements OAuth2ApiService, HasMetaStoreManagerFactory {
+@Named("default")
```
`@Named` has a couple drawbacks.
Using specialized `@Qualifier`s using `AnnotationLiteral`s to select the configured type via `Instance.select()` is the most flexible but also the safest way.

collado-mike replied:
Using `@Named` allows the existing configuration parsing to be used. HK2 does allow for accessing the different instances (e.g., using `IterableProvider`), but that's not really exclusive of using the `@Named` annotation

snazy replied:
There's `Instance.select()` in CDI for such cases - in combination with `@Qualifier` and `AnnotationLiteral` (mentioned elsewehere IIRC).

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/context/RealmScope.java:
```diff
@@ -16,12 +16,14 @@
-}
+@Scope
+@Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
```
```suggestion
@Target({ElementType.TYPE, ElementType.METHOD, ElementType.FIELD})
```

collado-mike replied:
hmm - right now, the bean scope is defined at either the factory or the bean class itself. I don't know that the dependents ought to decide the scope of the bean its depending on

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/context/RealmScope.java:
```diff
@@ -16,12 +16,14 @@
-
-  void setConfigurationStore(PolarisConfigurationStore configurationStore);
-}
+@Scope
```
```suggestion
@Scope
@Documented
```

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/context/RealmScope.java:
```diff
@@ -16,12 +16,14 @@
+@Scope
+@Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
+public @interface RealmScope {}
```
```suggestion
public @interface RealmScope extends AlterableContext {}
```

And have functionality to perform cleanup when a realm is removed.

collado-mike replied:
hmm, that's not in the jakarta cdi api jar

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:
```diff
@@ -157,12 +157,7 @@ public synchronized StorageCredentialCache getOrCreateStorageCredentialCache(
-  }
-
-  @Override
+  @Inject
```
Why not just add `@Inject` to the field?

collado-mike replied:
can do

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/persistence/PolarisEntityManager.java:
```diff
@@ -51,11 +51,14 @@ public class PolarisEntityManager {
    */
   public PolarisEntityManager(
-      PolarisMetaStoreManager metaStoreManager, StorageCredentialCache credentialCache) {
+      PolarisMetaStoreManager metaStoreManager,
```
Feels like `PolarisEntityManager` should become a CDI bean as well and get the instances injected.

collado-mike replied:
That would effectively eliminate the need for `RealmEntityManagerFactory` - which is fine, but it means the callers need to be made either realm-scoped or request-scoped

snazy replied:
Callers should be request-scoped, no?

adutra replied:
I asked myself this question as well when doing the Quarkus port. But in the end I left `PolarisEntityManager` as a non-CDI class. There might be room for improvement around here.

snazy commented on polaris-core/src/main/java/org/apache/polaris/core/persistence/cache/EntityCache.java:
```diff
@@ -34,6 +36,7 @@
 import org.jetbrains.annotations.Nullable;
 
 /** The entity cache, can be private or shared */
+@RealmScope
```
IMO a cache should leverage the available/configured resources of the JVM in a way that the resources are used most efficiently but also cannot exceed the available heap.

Here the cache is scoped to a realm, where each realm (number of realms is unbounded) permanently allocates JVM resources.

Not related to this PR, but something that should be fixed.

collado-mike replied:
That's a fine callout and I think revisiting the EntityCache to be realm-aware, rather than distinct per-realm, is a good change to make, but agreed that it shouldn't be done here

snazy commented on polaris-service/src/main/java/org/apache/polaris/service/BootstrapRealmsCommand.java:
```diff
@@ -48,17 +46,11 @@ protected void run(
       Namespace namespace,
       PolarisApplicationConfig configuration) {
-    MetaStoreManagerFactory metaStoreManagerFactory = configuration.getMetaStoreManagerFactory();
+    MetaStoreManagerFactory metaStoreManagerFactory =
```
Why not `@Inject` these?

collado-mike replied:
Dropwizard Commands aren't managed by HK2. We instantiate them directly in `PolarisApplication`.

snazy replied:
But that means that there's a third (CDI, Jackson, manual) way to get those instances. I think, CDI should be used everywhere (except for very specific tests).

adutra replied:
Commands (boostrap/purge) are going to be fun to adapt to CDI. I suggest we look into them later but my feeling is that we would need to move them to separate modules.

snazy commented on polaris-service/src/main/java/org/apache/polaris/service/config/PolarisApplicationConfig.java:
```diff
@@ -193,15 +348,15 @@ public long getMaxRequestBodyBytes() {
   }
 
-  public PolarisConfigurationStore getConfigurationStore() {
+  private PolarisConfigurationStore getConfigurationStore() {
```
Note: this is not a getter, but creates a new instance.

collado-mike replied:
yeah, I'll rename to be more clear

However, the factory is registered as a singleton, so the factory method is only invoked once

snazy commented on polaris-service/src/main/java/org/apache/polaris/service/persistence/InMemoryPolarisMetaStoreManagerFactory.java:
```diff
@@ -35,6 +37,8 @@
 import org.jetbrains.annotations.NotNull;
 
 @JsonTypeName("in-memory")
+@Named("in-memory")
```
Here there's CDI added, but also Jackson used as a CDI-ish mechanism.
In CDI, this is a `@Singleton`, but what about Jackson??

snazy commented on polaris-service/src/main/java/org/apache/polaris/service/tracing/TracingFilter.java:
```diff
@@ -53,6 +54,7 @@ public class TracingFilter implements Filter {
   private static final Logger LOGGER = LoggerFactory.getLogger(TracingFilter.class);
   private final OpenTelemetry openTelemetry;
 
+  @Inject
```
Generally, the CDI spec demands a public no-arg constructor - some frameworks do not require it though.

collado-mike replied:
The spec is pretty clear that you can annotate a constructor with its dependencies - https://jakarta.ee/specifications/cdi/3.0/jakarta-cdi-spec-3.0#bean_constructors

snazy commented on polaris-service/src/main/resources/META-INF/hk2-locator/default:
```diff
@@ -0,0 +1,58 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+[org.apache.polaris.service.auth.DefaultPolarisAuthenticator]S
```
Feels like this hk2 specific way to provide instances should be replaced with CDI producers?

snazy commented on regtests/Dockerfile:
```diff
@@ -54,4 +54,4 @@ USER root
 USER spark
 
-CMD ["./run.sh"]
+ENTRYPOINT ["./run.sh"]
```
Change to this PR?

collado-mike replied:
added when I was debugging this test failure- https://github.com/apache/polaris/pull/493/commits/e5367d96273bf0560f0ce5bef8b003c1e20ffeeb . TBH, I'm not sure how tests in main were passing

snazy reviewed (commented):
Thanks for the nice improvement towards CDI!

hk2 adds a DI implementation, but in a couple of places this conflicts with the way Jackson is used as a DI-ish mechanism to instantiate things, leading to multiple DI-(ish) implementations.

I'm not sure how an approach to add "generic" extensions/plugins would look with hk2, as we cannot expect to have initialization code for all extensions/plugins in the "main" code base.

dimas-b commented on polaris-service/src/main/java/org/apache/polaris/service/config/PolarisApplicationConfig.java:
```diff
@@ -63,69 +88,206 @@ public class PolarisApplicationConfig extends Configuration {
+    public void dispose(T instance) {}
+  }
+
+  public <T> T findService(Class<T> serviceClass) {
```
This seems to be a CDI concern rather than a configuration concern. I mean that it is perfectly fine to drive CDI with config settings, but `findService` does not appear to fit the DI paradigm :thinking:

collado-mike replied:
It doesn't, but the servlet filters don't use DI, so the PolarisApplication has to find them somehow

dimas-b commented on polaris-service/src/main/resources/META-INF/hk2-locator/default:
```diff
@@ -0,0 +1,88 @@
+name=default
+qualifier={jakarta.inject.Named}
+
+[org.apache.polaris.service.auth.TestInlineBearerTokenPolarisAuthenticator]S
+contract={io.dropwizard.auth.Authenticator}
+name=test
+qualifier={jakarta.inject.Named}
```
I'm not sure I understand the purpose of this declaration. It seems rather redundant to me. The class of `TestInlineBearerTokenPolarisAuthenticator` already has the `@Named("test")` annotation and implements `Authenticator`. This file does not appear to add any new information to what is already known from class files :thinking:

collado-mike replied:
This is something that would normally be generated at build time by scanning the build artifacts. But that requires using the HK2-specific annotations (`@Contract` and `@Service`), which I think we agreed to avoid

snazy commented:
Hm, looking into the [hk2 github repo](https://github.com/eclipse-ee4j/glassfish-hk2) it looks like its rather unmaintained (minus version bumps and rather tiny changes).

dimas-b commented:
I'd also be willing to work on extracting services (generated API code + impl. classes) into a dedicated gradle module (to be pulled into DW runtime), unless you want to work on this from your side @collado-mike :)

dimas-b commented on polaris-service/src/main/java/org/apache/polaris/service/context/RealmScopeContext.java:
```diff
@@ -0,0 +1,82 @@
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <U> U findOrCreate(ActiveDescriptor<U> activeDescriptor, ServiceHandle<?> root) {
```
Is it possible to implement `@RealmScope` as a `@RequestScope` with custom bean producers that handle the realm lookup and caching? I think that might be more intuitive from the CDI perspective... and probably more portable. WDYT?

collado-mike replied:
I'd worry that the intention was not clear - readers would just have to know that an item annotated with `@RequestScoped` could actually live beyond the scope of a single request. In fact, they could be used by multiple concurrent requests, while the `@RequestScoped` annotation would convey that it's safe for an object to mutate state possibly leading to thread safety errors. I think all of the CDI providers offer support for custom scopes and it doesn't seem like a lot of work for what amounts to (IMO) clear intentions.

adutra replied:
I think we are mixing concepts here, a request scoped bean should not outlive the request. In the Quarkus PR the `RealmContext` is a request-scoped bean, and one instance of it is created for each request. It two requests target the same realm, they will get two distinct instances of `RealmContext`, but each would have the same realm ID.

collado-mike replied:
We're specifically talking about beans that should be reused across requests

dimas-b replied:
Why not have a global store for realm-specific stuff and expose it to callers via thin request-scoped wrappers? Disposing of the wrapper does not have to dispose of the long-lived data.

... besides, we never destroy realm-scoped beans anyway.

collado-mike replied:
It seems to me that's exactly what the `@RealmScoped` annotation is doing. The `RealmScopeContext` maintains the cache per-realm and manages the instantiated values for each `ActiveDescriptor` for the currently active realm. This is a centralized place where we can do a lot of cleanup of stale realms (e.g., somebody made a call two days ago, but hasn't invoked an API since), rather than having to sprinkle this cache eviction in a bunch of different bean-specific factories. 

I think the contention is whether we should be relying on a ThreadLocal to access the current `RealmContext`. I can see where there are flaws with this approach (e.g., we already ran into one during the asynchronous cleanup of metadata files after table drop), but at the same time, I think most (every?) implementation of `@RequestScoped` relies on ThreadLocal variables (e.g., AFAICT, ArC relies on [ThreadLocalCurrentContext](https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ThreadLocalCurrentContext.java) in its [RequestContext](https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java#L66) ). Injecting an accessor to retrieve the current realm makes a lot of sense to me, but at the end of the day, I think we're still going to be relying on ThreadLocals.

collado-mike commented:
> I'd also be willing to work on extracting services (generated API code + impl. classes) into a dedicated gradle module (to be pulled into DW runtime), unless you want to work on this from your side @collado-mike :)

That sounds good to me! I have other things to work on :)

dimas-b commented on polaris-core/src/main/java/org/apache/polaris/core/context/RealmScope.java:
```diff
@@ -16,12 +16,16 @@
+@Documented
+@Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
+public @interface RealmScope {}
```
nit: rename to `RealmScoped` similar to `@RequestScoped`?

adutra commented on polaris-service/src/main/java/org/apache/polaris/service/context/RealmScopeContext.java:
```diff
@@ -0,0 +1,82 @@
+
+  @Override
+  public boolean isActive() {
+    return CallContext.getCurrentContext() != null
```
Here we are using `CallContext.getCurrentContext()`, which internally relies on thread-locals. This is NOT going to work well with CDI. We have already opened an issue to raise awareness of that: #463. My long-term vision for `CallContext` and alike is:

* `CallContext` and `PolarisCallContext` must disappear eventually;
* `RealmContext` can stay, and become request-scoped.

In the meanwhile, since removing `CallContext` is a large change, what I did in the Quarkus PR is that I turned all 3 beans into request-scoped, and they are `@Inject`-ed whenever required, avoiding `CallContext.getCurrentContext()` as much as possible.

collado-mike replied:
This has been updated to rely on the `RealmContext` bean's scope. Whether or not the `CallContext` remains is a discussion I think we should have on a future PR

adutra commented on polaris-service/src/main/java/org/apache/polaris/service/context/RealmScopeContext.java:
```diff
@@ -0,0 +1,82 @@
+import org.glassfish.hk2.api.ServiceHandle;
+
+@Singleton
+public class RealmScopeContext implements Context<RealmScope> {
```
So, I'm really, really not comfortable with this `RealmScopeContext`. 

Speaking from experience, most custom scopes are hard to maintain, and can introduce subtle bugs that are hard to chase. For instance, it's never easy to know when to close a custom scope.

OTOH, most custom scoped-beans can actually be retrofitted to be request-scoped beans with a bit of refactoring. This is imho much safer because we don't have to care about when to open or when to close the scope.

In the Quarkus PR, `RealmContext` is a request-scoped bean, and that's working fine.

collado-mike replied:
The concern is not whether the `RealmContext` is recreated, but the items that are supposed to be scoped to a realm and reused across requests. Some examples off the top of my head are
* EntityCache
* Threadpools for background work
* HTTP clients for remote caches or token validation
* DB Connection pools

There are a lot of good reasons to want to reuse resources across requests (e.g., https connections are expensive and we want to reuse them as much as possible), but it's important that these resources are scoped to specific realms (we don't want to look up `server_admin` from one realm when processing a request for another).

I do get the concern about subtle bugs, though. That's one reason I added an explicit test to validate the entity cache is scoped to the realm and reused across requests. More tests and validation would be good for comfort.

We could wrap all the realm-scoped stuff in factories with hash maps, like the `MetaStoreManagerFactory`, but... it just means we have to write factories for all those things. And then, as I commented to Dmitri, the intention becomes less clear - if readers believe a bean is request-scoped, but is actually supplied from a cache and reused across requests, it may be unclear if the object is safe for concurrent modification.

collado-mike replied:
This particular `RealmScopeContext` is HK2-specific and won't be portable to Quarkus. The annotation, however, can be. I experimented with this briefly with another CDI impl and I was able to write basically the same code without relying on thread-locals (except as used by the underlying framework). Maybe we can chat about the specific impl in the Quarkus PR? For now, I think we get the functionality we want without the need to add a bunch of new factories to this PR and there's nothing in the core logic that ties anything to HK2 in particular

adutra commented on polaris-service/src/main/java/org/apache/polaris/service/persistence/InMemoryPolarisMetaStoreManagerFactory.java:
```diff
@@ -35,9 +35,10 @@
 
-@JsonTypeName("in-memory")
+@Named("in-memory")
+@Singleton
```
We have to decide between `@Singleton` and `@ApplicationScoped` – unless we have a good reason to use the former, the latter is preferred.

collado-mike replied:
`@ApplicationScoped` is not part of the jakarta inject-api (it's in the enterprise package), so isn't supported by Dropwizard/HK2 today. I'm fine to remove this annotation and keep the scope defined in the configuration, but bringing in the jakarta enterprise-inject-api is a lot more work and I think we should keep this PR scoped

adutra replied:
Let's remove then, and let the CDI runtime decide what the scope is.

adutra commented on polaris-service/src/main/java/org/apache/polaris/service/persistence/cache/EntityCacheFactory.java:
```diff
@@ -0,0 +1,47 @@
+  public EntityCache provide() {
+    LOGGER.debug(
+        "Creating new EntityCache instance for realm {}",
+        CallContext.getCurrentContext().getRealmContext().getRealmIdentifier());
```
`RealmContext` should be injected, not retrieved from thread locals. That's what `RealmEntityManagerFactory` does, cf. `getOrCreateEntityManager(RealmContext)`.

collado-mike replied:
I think that's a great suggestion

adutra commented on polaris-core/src/main/java/org/apache/polaris/core/persistence/LocalPolarisMetaStoreManagerFactory.java:
```diff
@@ -54,7 +54,7 @@ public abstract class LocalPolarisMetaStoreManagerFactory<StoreType>
   protected final PolarisDiagnostics diagServices = new PolarisDefaultDiagServiceImpl();
 
-  protected PolarisStorageIntegrationProvider storageIntegration;
+  @Inject protected PolarisStorageIntegrationProvider storageIntegration;
```
This should be removed from here (along with the setter below) because this field is not used in this class.

Instead it should be injected on concrete subtypes that actually need it.

adutra commented on polaris-service/src/main/java/org/apache/polaris/service/config/RealmEntityManagerFactory.java:
```diff
@@ -18,29 +18,39 @@
 
 /** Gets or creates PolarisEntityManager instances based on config values and RealmContext. */
+@RealmScope
 public class RealmEntityManagerFactory {
```
I disagree here, this is clearly an application-scoped bean, holding long-lived instances of `PolarisEntityManager` keyed by realm id.

collado-mike replied:
You're right. I'd actually like to get rid of this, as I don't think it adds much with the CDI infra, but I'll keep it as a singleton for now

adutra commented on polaris-service/src/main/java/org/apache/polaris/service/context/DefaultContextResolver.java:
```diff
@@ -166,7 +165,7 @@ private static Map<String, String> parseBearerTokenAsKvPairs(Map<String, String>
   }
 
-  @Override
+  @Inject
```
setter -> constructor

adutra commented on polaris-service/src/main/java/org/apache/polaris/service/persistence/cache/EntityCacheFactory.java:
```diff
@@ -0,0 +1,47 @@
+  private static Logger LOGGER = LoggerFactory.getLogger(EntityCacheFactory.class);
+  @Inject PolarisMetaStoreManager metaStoreManager;
+
+  @RealmScope
```
Again I think it's much easier to turn this bean into application-scoped and have it hold a map of entity caches keyed by realm id, like `RealmEntityManagerFactory`.

adutra commented on polaris-core/src/main/java/org/apache/polaris/core/storage/PolarisStorageIntegrationProviderImpl.java:
```diff
@@ -16,29 +16,26 @@
  * under the License.
  */
-package org.apache.polaris.service.storage;
+package org.apache.polaris.core.storage;
```
Why move an implementation class to core?

collado-mike replied:
This was leftover from a previous attempt. I changed it back.

adutra commented on server-templates/api.mustache:
```diff
@@ -83,6 +85,7 @@ public class {{classname}}  {
 
   private final {{classname}}Service service;
 
+  @Inject
```
If you turn `RealmContext` into request-scoped, a nice benefit that comes for free is that you could inject it directly into the REST resources generated from the open-api spec. This way, the realm is immediately available and injected everywhere. E.g. the /config endpoint could become:

```java
  public Response getConfig(
      @QueryParam("warehouse") String warehouse,
      @Context RealmContext realmContext,
      @Context SecurityContext securityContext) {
    return service.getConfig(warehouse, realmContext, securityContext);
  }
```

collado-mike commented:
> I agree that @Named has some shortcomings.
> One of them is that the name is application-wide, so you cannot have two beans named @Named("default").

@adutra I haven't seen this behavior during testing. E.g., the `DefaultOAuth2ApiService` and the `DefaultContextResolver` are both named "default", but the application doesn't seem to have any problem starting up both instances. Could it be that we never use `@Named` as a qualifier at the injection point? From the article you shared, it seems like the issue is one of ambiguity at injection time if the qualifiers aren't specific. But the goal of the `@Named` annotation in this PR is to allow named instances to be specified in the configuration. We never use the `@Named` qualifier at the injection points (and I don't think we should, as it defeats the purpose of the user configuration).

collado-mike commented on polaris-service/src/main/java/org/apache/polaris/service/auth/DefaultPolarisAuthenticator.java:
```diff
@@ -36,16 +36,7 @@ public Optional<AuthenticatedPolarisPrincipal> authenticate(String credentials)
-  }
-
-  @JsonProperty("tokenBroker")
+  @Inject
```
Objects that are constructed during the configuration parsing have to have no-arg constructors. We can do field or setter injection, but not constructor injection for these types. This includes the `Authenticator`, the `TokenBroker`, the `MetaStoreManagerFactory` and all other types defined in the `PolarisApplicationConfig`.

adutra replied:
Hmm okay. Can we go with field injection then?

collado-mike commented on polaris-service/src/main/java/org/apache/polaris/service/auth/JWTRSAKeyPairFactory.java:
```diff
@@ -39,7 +39,7 @@ public TokenBroker apply(RealmContext realmContext) {
   }
 
-  @Override
+  @Inject
```
See above response

adutra commented:
> @adutra I haven't seen this behavior during testing. E.g., the `DefaultOAuth2ApiService` and the `DefaultContextResolver` are both named "default", but the application doesn't seem to have any problem starting up both instances.

I suspect HK2 might not be fully compliant with the spec. From [CDI spec 4.0](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#ambig_names):

> An ambiguous name exists when a name resolves to multiple beans. When an ambiguous name exists, the container attempts to resolve the ambiguity.

So in our case, if both `DefaultOAuth2ApiService` and `DefaultContextResolver` are named `default`, and are both eligible for injection, then that should result in an unresolvable ambiguous bean situation.

For example with Quarkus, if I declare the following beans:

```java
@Singleton @Named("foo")
public class Foo1 {}

@Singleton @Named("foo")
public class Foo2 {}
```

I get a deployment error when the app starts (even if the beans aren't injected anywhere):

```
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: jakarta.enterprise.inject.spi.DeploymentException: Unresolvable ambiguous bean name detected: foo
Beans:
CLASS bean [types=[org.apache.polaris.service.catalog.Foo1, java.lang.Object], qualifiers=[@Default, @Any, @Named("foo")], target=org.apache.polaris.service.catalog.Foo1]
CLASS bean [types=[org.apache.polaris.service.catalog.Foo2, java.lang.Object], qualifiers=[@Default, @Any, @Named("foo")], target=org.apache.polaris.service.catalog.Foo2]
```

To get past this issue, I would suggest that we copy [SmallRye's `@Identifier`](https://github.com/smallrye/smallrye-common/blob/main/annotation/src/main/java/io/smallrye/common/annotation/Identifier.java) into our code base.

collado-mike commented:
> To get past this issue, I would suggest that we copy [SmallRye's `@Identifier`](https://github.com/smallrye/smallrye-common/blob/main/annotation/src/main/java/io/smallrye/common/annotation/Identifier.java) into our code base – if it's possible for HK2 to process custom qualifiers.

Yeah, it's possible to use the custom qualifiers. I think it's fine to switch the annotations - but copying the `@Identifier` code will make the classes that use it incompatible with the Quarkus work, right? I think it would be fine to just add a dependency on the SmallRye library for the annotation, as that moves the codebase into a more compatible state. WDYT?

adutra commented:
> Yeah, it's possible to use the custom qualifiers. I think it's fine to switch the annotations - but copying the `@Identifier` code will make the classes that use it incompatible with the Quarkus work, right? I think it would be fine to just add a dependency on the SmallRye library for the annotation, as that moves the codebase into a more compatible state. WDYT?

Copying wouldn't break Quarkus afaict, it's just a matter of telling it which identifier we intend use. That said, if you are OK adding the dependency, that's even better 👍

collado-mike commented:
@adutra , I think I've addressed all of your comments. The one outstanding concern is the support for the `@RealmScoped` annotation. I think we should discuss the feasibility of supporting that in Quarkus in the Quarkus PR. Its use and the context to support it are framework-dependent anyway, so I think we can evaluate the feasibility of it specifically in Quarkus separately from this PR.

In the meantime, however, I think the basic CDI support is blocking work on several other tasks (e.g., the authentication/authorization PRs Dmitri and I have open, fixing the bootstrapping, table metadata logging). With this PR, I think we can work on better/more complete CDI support (e.g., through Quarkus) without any significant changes to the work done here and it can also accommodate the other work that's ongoing.

adutra reviewed (approved):
I'm approving in the spirit of allowing us to move forward. There will be issues to solve for sure, but I think we can fix them along the way. 👍

collado-mike commented:
@adutra , I'll need another approval because I had to merge from main to resolve conflicts 🙄
